src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java

```diff
@@ -125,16 +237,17 @@ public void newDirectoryStream() throws ExecutionException, InterruptedException
         when(mockClient.listObjectsV2(any(ListObjectsV2Request.class))).thenReturn(CompletableFuture.supplyAsync(() ->
                 ListObjectsV2Response.builder().contents(object1, object2).build()));
 
-        final DirectoryStream<Path> stream = provider.newDirectoryStream(mockClient, Paths.get(URI.create(pathUri+"/")), entry -> true);
+        final DirectoryStream<Path> stream = provider.newDirectoryStream(Paths.get(URI.create(pathUri+"/")), entry -> true);
         assertNotNull(stream);
         assertEquals(2, countDirStreamItems(stream));
     }
 
-    /**
-     * Tests core functionality of the path iterator used by the directory stream.
-     */
     @Test
-    public void pathIteratorForPublisher() {
+    @Disabled
```

Comment:
I'm not comfortable disabling this. If the pagination is not working then we have a regression and should investigate.

Reply:
I agree 100%. As per comment https://github.com/awslabs/aws-java-nio-spi-for-s3/pull/133#discussion_r1285169401 I need your help. I think I found a clean solution to make the iterator private, but I could not make it work the test. I believe it is my lack of familiarity with mocking S3 objects because the logic is correct:

`provider.newDirectoryStream(Paths.get(URI.create(pathUri+"/")), filter).iterator()`

DirectoryStream already provides a way to get the iterator. Once you get that, the test should work exactly like before. Can you please have a look and help me out with that?

Reply:
The challenge with making it private is that you can't provide it with an S3Client which ultimately means you can't provide it with a mock S3Client. So the Directory iterator you get will be using a regular S3Client and will try to call the S3 service which will fail. 

One way you could do it is to make a separate protected or even public ListObjectsV2Provider and create one in the constructor of the appropriate class as a private variable.

Reply:
I do not think so, @markjschreiber. If you look run the test, what fails is the last test only, which means to me if is using the mock passed in in creating the publisher

https://github.com/stefanofornari/aws-java-nio-spi-for-s3/blob/896d1e812405a23bcfb007f134e1e7474bc2b45d/src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java#L251 

Plus, the test just before 

https://github.com/stefanofornari/aws-java-nio-spi-for-s3/blob/896d1e812405a23bcfb007f134e1e7474bc2b45d/src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java#L225C24-L225C24

Which basically the same, but it does not paginate, works, therefore, no it can not be that. The thing is that pagination is not triggered and maybe it was not before neither. I tend to believe it is in the way the mock is prepared. How can we trigger pagination by a proper client response?

Reply:
I will keep investigating it, but since you developed that piece, I am pretty sure that you can spot the issue quickly if you give it a run.

Reply:
Ok, I've got it... it was actually a bug

The problem was that in here:

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/98e2f9da844b12443b1841373a0643a4e56c576a/src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java#L348
p is a string containing the full path like s3://foo/baa/, while finalDirName contains only the path 'baa/' because it comes from:

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/98e2f9da844b12443b1841373a0643a4e56c576a/src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java#L303

where finalDirName is the result of s3Path.toAbsolutePath().getKey(); and the two never match.

The reason why your test was passing is because you gave a matching string in your case (you pass in pathUri+"/" as string argument, which turns into "s3://foo/baa/"), but this is not what production code does:

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/98e2f9da844b12443b1841373a0643a4e56c576a/src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java#L150-L154

This is also why I avoid using mocks as much as I can ;)

I am updating the PR with a possible fix, please let me know if it is fine,